app/src/hvr/java/com/igalia/wolvic/HVRSpeechRecognizer.java

```diff
@@ -6,21 +6,58 @@
 import android.os.Bundle;
 
 import androidx.annotation.NonNull;
-import androidx.annotation.Nullable;
 
 import com.huawei.hms.mlsdk.asr.MLAsrConstants;
 import com.huawei.hms.mlsdk.asr.MLAsrListener;
 import com.huawei.hms.mlsdk.asr.MLAsrRecognizer;
-
 import com.igalia.wolvic.speech.SpeechRecognizer;
+import com.igalia.wolvic.utils.LocaleUtils;
+import com.igalia.wolvic.utils.StringUtils;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.List;
+import java.util.Locale;
 
 public class HVRSpeechRecognizer implements SpeechRecognizer, MLAsrListener {
     private Context mContext;
     private MLAsrRecognizer mRecognizer;
     private Callback mCallback;
 
+    // Language support depends on the region and English is the only one supported everywhere.
+    // https://developer.huawei.com/consumer/en/doc/development/hiai-Guides/ml-asr-0000001050066212
+    private static final List<String> DEFAULT_SUPPORTED_LANGUAGES = Collections.singletonList(MLAsrConstants.LAN_EN_US);
+    // The first element in these two lists corresponds to "Follow device language"
+    private final List<String> mSupportedLanguages = new ArrayList<>();
+    private final List<String> mSupportedLanguageNames = new ArrayList<>();
+
     public HVRSpeechRecognizer(Context context) {
         mContext = context;
+
+        setSupportedLanguages(DEFAULT_SUPPORTED_LANGUAGES);
+        mRecognizer = MLAsrRecognizer.createAsrRecognizer(mContext);
+        mRecognizer.getLanguages(new MLAsrRecognizer.LanguageCallback() {
+            @Override
+            public void onResult(List<String> list) {
+                setSupportedLanguages(list);
+            }
+
+            @Override
+            public void onError(int i, String s) {
+            }
+        });
+    }
+
+    private synchronized void setSupportedLanguages(List<String> supportedLanguages) {
```

Comment:
This is only called from the constructor, why do we need to make it synchronized?

Reply:
I was trying to avoid a possible error because fetching the list of languages happens asynchronously. However, it doesn't look possible to have a concurrency problem because the call to fetch the languages is only invoked once, in the constructor.

I will remove this.